agent/ebs/watcher.go

```diff
@@ -0,0 +1,212 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ebs
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+	apiebs "github.com/aws/amazon-ecs-agent/ecs-agent/api/resource"
+	log "github.com/cihub/seelog"
+)
+
+type EBSWatcher struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	scanTicker *time.Ticker
+	agentState dockerstate.TaskEngineState
+	// TODO: The dataClient will be used to save to agent's data client as well as start the ACK timer. This will be added once the data client functionality have been added
+	// dataClient     data.Client
+	ebsChangeEvent  chan<- statechange.Event
+	discoveryClient apiebs.EBSDiscovery
+	mailbox         chan func()
+}
+
+// NewWatcher is used to return a new instance of the EBSWatcher struct
+func NewWatcher(ctx context.Context,
+	state dockerstate.TaskEngineState,
+	stateChangeEvents chan<- statechange.Event) (*EBSWatcher, error) {
+	derivedContext, cancel := context.WithCancel(ctx)
+	discoveryClient := apiebs.NewDiscoveryClient(derivedContext)
+	return &EBSWatcher{
+		ctx:             derivedContext,
+		cancel:          cancel,
+		agentState:      state,
+		ebsChangeEvent:  stateChangeEvents,
+		discoveryClient: discoveryClient,
+		mailbox:         make(chan func(), 100),
+	}, nil
+}
+
+// Start is used to kick off the periodic scanning process of the EBS volume attachments for the EBS watcher.
+// If there aren't any initially, the scan ticker will stop.
+func (w *EBSWatcher) Start() {
+	log.Info("Starting EBS watcher.")
+
+	w.scanTicker = time.NewTicker(apiebs.ScanPeriod)
+	if len(w.agentState.GetAllPendingEBSAttachments()) == 0 {
+		w.scanTicker.Stop()
```

Comment:
When is this  EBSWatcher `Start()` called?  during resource initialization?

Reply:
Great questions. We'll only have one ticker active at all times. The EBSWatcher `Start()` will be called whenever we have a pending EBS attachment to find (i.e. whenever we receive an attachment message). It'll also start on agent start up (will be done in a very near PR). 

Note: It will start up whenever we have a list of pending EBS volumes of size 1.